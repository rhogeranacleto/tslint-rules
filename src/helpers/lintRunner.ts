import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}