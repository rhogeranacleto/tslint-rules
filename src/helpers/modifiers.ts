export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}