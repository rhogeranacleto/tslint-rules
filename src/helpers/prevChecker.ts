// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}