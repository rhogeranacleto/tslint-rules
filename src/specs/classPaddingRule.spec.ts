import { lintRunner } from '../helpers/lintRunner';

const RULE = 'class-padding';

describe('ClassPadding', () => {

	it.each([
		[`class Maria {
			constructor(private jj: JJ){}
			public dan() {
				console.log('oi');
			}
			public dan() {
				console.log('oi');
			}
		}`, 1],
		[`class Maria {

			public dan() {
				console.log('oi');
			}

		}`, 1],
		[`class Maria {
			public dan() {
				console.log('oi');
			}

		}`, 2],
		[`class Maria {

			constructor(private jj: JJ) {}

			public dan() {
				console.log('oi');
			}
		}`, 0],
		[`class Maria {

		}`, 1],
		[`@Entity()
		class Maria {

		}`, 1],
		[`class Maria {
		}`, 1],
		[`class Maria { }`, 0],
		[`@Entity()
		class Maria { }`, 0],
		[`@Injectable()
		export class SubSystemService {

			constructor(private readonly benchmarkingValidationsService: BenchmarkingValidationsService) { }

			public async edit(id: number, subSystem: SubSystem) {

				if (!subSystem.in_revision) {

					await Promise.all([
						this.benchmarkingValidationsService.validateInRevisionAssociations(SubSystem, subSystem),
						this.benchmarkingValidationsService.validateEquivalentRecordAssociations(SubSystem, subSystem, id)
					]);
				}

				return this.repository.updateOne(id, subSystem);
			}
		}`, 0]
	])('%s', (src, count) => {

		const result = lintRunner({ src, rule: RULE });

		expect(result.errorCount).toBe(count);
	});

	it.each([
		[`import jj from 'jj';
		class Maria {}`, 1],
		[`import jj from 'jj';
		@Entity()
		class Maria {}`, 1],
		[`import jj from 'jj';

		@Entity()
		class Maria {}`, 0],
		[`import jj from 'jj';

		class Maria {}`, 0]
	])('%s', (src, count) => {

		const result = lintRunner({ src, rule: RULE });

		expect(result.errorCount).toBe(count);
	});
});