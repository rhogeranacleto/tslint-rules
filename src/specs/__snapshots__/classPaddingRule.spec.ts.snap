// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassPadding @Entity()
		class Maria {

		} 1`] = `
Array [
  Array [
    "@Entity()
		class Maria { }",
  ],
]
`;

exports[`ClassPadding @Entity()
		class SemMaria {

		} 1`] = `
"@Entity()
		class SemMaria { }"
`;

exports[`ClassPadding class Maria {
			constructor(private jj: JJ){}
			public dan() {
				console.log('oi');
			}
			public dan() {
				console.log('oi');
			}
		} 1`] = `
"class Maria {

			constructor(private jj: JJ){}
			public dan() {
				console.log('oi');
			}
			public dan() {
				console.log('oi');
			}
		}"
`;

exports[`ClassPadding class Maria {
		} 1`] = `
Array [
  Array [
    "class Maria { }",
  ],
]
`;

exports[`ClassPadding class Maria {

			public dan() {
				console.log('oi');
			}

		} 1`] = `
"class Maria {

			public dan() {
				console.log('oi');
			}
		}"
`;

exports[`ClassPadding class Maria {

		} 1`] = `
Array [
  Array [
    "class Maria { }",
  ],
]
`;

exports[`ClassPadding class OtraMaria {
		} 1`] = `"class OtraMaria { }"`;

exports[`ClassPadding class TaMaria {

		} 1`] = `"class TaMaria { }"`;

exports[`ClassPadding import jj from 'jj';
		@Entity()
		class Maria {} 1`] = `
"import jj from 'jj';

		@Entity()
		class Maria {}"
`;

exports[`ClassPadding import jj from 'jj';
		class Maria {} 1`] = `
"import jj from 'jj';

		class Maria {}"
`;
